src/mds/Locker.cc

```diff
@@ -225,6 +225,41 @@ struct MarkEventOnDestruct {
   }
 };
 
+/*
+ * Try to detect the possible deadlock between requests, and
+ * will try to break the newer request.
+ */
+void Locker::detect_dead_locks(const MDRequestRef& mdr, int lock_type)
+{
+  bool any_waiter = false;
+  bool order_mess = false;
+
+  if (!g_conf().get_val<bool>("mds_detect_deadlocks")) {
+    return;
+  }
+
+  dout(15) << "detect_dead_locks " << *mdr << dendl;
+  for (auto it = mdr->locks.begin(); it != mdr->locks.end(); ) {
+    SimpleLock *lock = it->lock;
+
+    if (lock->is_waiter_for(SimpleLock::WAIT_ALL)) {
+      any_waiter = true;
+    }
+    if (lock->get_type() > lock_type) {
+      dout(1) << " detected mess lock order, locks: " << mdr->locks << dendl;
+      order_mess = true;
+    }
```

Comment:
(just trying to understand locks better), so this is the list i see in ceph_fs.h
```
#define CEPH_LOCK_DN          (1 << 0)
#define CEPH_LOCK_DVERSION    (1 << 1)
#define CEPH_LOCK_IQUIESCE    (1 << 4)  /* mds internal */
#define CEPH_LOCK_ISNAP       (1 << 5)  /* snapshot lock. MDS internal */
#define CEPH_LOCK_IPOLICY     (1 << 6)  /* policy lock on dirs. MDS internal */
#define CEPH_LOCK_IFILE       (1 << 7)
#define CEPH_LOCK_INEST       (1 << 8)  /* mds internal */
#define CEPH_LOCK_IDFT        (1 << 9)  /* dir frag tree */
#define CEPH_LOCK_IAUTH       (1 << 10)
#define CEPH_LOCK_ILINK       (1 << 11)
#define CEPH_LOCK_IXATTR      (1 << 12)
#define CEPH_LOCK_IFLOCK      (1 << 13)  /* advisory file locks */
#define CEPH_LOCK_IVERSION    (1 << 14)  /* mds internal */
```

when we say `lock->get_type() > lock_type` is a mess, does it mean that the locks less than the `lock->get_type()` would be acquired?

Reply:
@dparmar18 when multiple locks need to be taken, they should be taken in a pre-defined order. If that order will be observed by all concurrent lockers, no deadlocks will occur.

CephFS approach to this is to sort locks in the `acquire_locks` by the lock_type:

```
bool Locker::acquire_locks(const MDRequestRef& mdr,
			   MutationImpl::LockOpVec& lov,
 ...
{
  ...
  lov.sort_and_merge();
  ...
```

This would be sufficient if all requests always specified the full list of locks they need upfront, but that's not the case, and in many instances would be counter-productive.

Some flows choose to start by taking a subset of locks and then add more locks as they progress. Such flows must be careful, because if they began by only taking, say, `xattr` lock, and then they want to take the `file` lock, they would violate the said lock order, and `acquire_locks` wouldn't be able to help here, as separate calls to the method mention the locks.
Such multi-phase flows should take the responsibility to only acquire the locks in the correct order, and if then need to violate it, then they have to drop locks that they already hold before attempting to acquire a lock with a lower `lock_type`. In my example, the request would have to drop the xattr lock and then acquire both file and xattr lock in a single call to maintain the order and prevent potential deadlocks.
It's not always possible to do, because dropping xattr lock would break atomicity of the flow wrt xattr values, which may or may not be acceptable. If we can't afford it, the solution for that example flow would be to take file lock along with xattr lock from the beginning, for atomicity.

Reply:
Hey @leonid-s-usov thanks for taking time on this, couple of questions here:

> This would be sufficient if all requests always specified the full list of locks they needed upfront, but that's not the case, and in many instances, it would be counter-productive.
> 

Why/how could it be counter-productive? An already known lock set would mean better clarity?

> Some flows start by taking a subset of locks and later add more locks as they progress. Such flows must be careful, because if they began by only taking, say, `xattr` lock, and then they want to take the `file` lock, they would violate the said lock order, and `acquire_locks` wouldn't be able to help here, as separate calls to the method mention the locks.

> 
> Such multi-phase flows should take the responsibility to only acquire the locks in the correct order, and if they need to violate it by taking new locks, then they have to drop any locks that they already hold with the `lock_type` greater than the lowest `lock_type` of the new locks they need. 

- As you mentioned earlier some code paths start with a sub-set and then add locks as per their need, till now (since this patch's need came up recently) I assume things worked fine before? and maybe the quesce protocol added broke things up? or any new stuff added?
- Shouldn't `LockOpVec` be used to make sure there are no such deadlocks? Rationale for this statement: it is a seq of the lock ops to be performed to ensure consistency, right? so if we're now doube-checking this then it means we aren't confident of the seq in `LockOpVec`? i.e. we aren't sure if the seq in the `LockOpVec` is in pre-defined order? shouldn't the `LockOpVec` be having a valid seq then? or am i misunderstanding the use case of `LockOpVec` and it is not that smart enough?
- If we're dropping locks just like this, won't it be disastrous? There can be some I/O, we preempt it, the other thread/process that should've waited is now having the lock and the process earlier has some unfinished work left, doesn't this lead to data corruption/inconsistencies/loss?

>In my example, the request would have to drop the `xattr` lock and then acquire both the `file` and the `xattr` locks in a single call to maintain the order and prevent potential deadlocks. It's not always possible to do, because dropping xattr lock would break the atomicity of the flow wrt the values protected by the xattr lock, which may or may not be acceptable. If we can't afford it, the solution for that example flow would be to take file lock along with xattr lock from the beginning, for atomicity.

so lets say a req had a xattr lock and now wants a file locks, so the MDS would drop the xattr lock then acquire the file lock and then xattr lock? and in the meantime what happens if the something else took up xattr?

Reply:
> Hey @leonid-s-usov thanks for taking time on this, couple of questions here:
> 
> > This would be sufficient if all requests always specified the full list of locks they needed upfront, but that's not the case, and in many instances, it would be counter-productive.
> 
> Why/how could it be counter-productive? An already known lock set would mean better clarity?
> 
> > Some flows start by taking a subset of locks and later add more locks as they progress. Such flows must be careful, because if they began by only taking, say, `xattr` lock, and then they want to take the `file` lock, they would violate the said lock order, and `acquire_locks` wouldn't be able to help here, as separate calls to the method mention the locks.
> 
> > Such multi-phase flows should take the responsibility to only acquire the locks in the correct order, and if they need to violate it by taking new locks, then they have to drop any locks that they already hold with the `lock_type` greater than the lowest `lock_type` of the new locks they need.
> 
> * As you mentioned earlier some code paths start with a sub-set and then add locks as per their need, till now (since this patch's need came up recently) I assume things worked fine before? and maybe the quesce protocol added broke things up? or any new stuff added?

Mostly it worked well before, but recently we have hit at least two bugs caused by the lock order issue, please see https://github.com/ceph/ceph/pull/52522 and https://tracker.ceph.com/issues/65607.

Not introduced by the quesce changes.

> * Shouldn't `LockOpVec` be used to make sure there are no such deadlocks? Rationale for this statement: it is a seq of the lock ops to be performed to ensure consistency, right? so if we're now doube-checking this then it means we aren't confident of the seq in `LockOpVec`? i.e. we aren't sure if the seq in the `LockOpVec` is in pre-defined order? shouldn't the `LockOpVec` be having a valid seq then? or am i misunderstanding the use case of `LockOpVec` and it is not that smart enough?

The `LockOpVec` can only guarantee the lock order in each `acquire_locks()`, but couldn't guarantee multiple `acquire_locks()`.

> * If we're dropping locks just like this, won't it be disastrous? There can be some I/O, we preempt it, the other thread/process that should've waited is now having the lock and the process earlier has some unfinished work left, doesn't this lead to data corruption/inconsistencies/loss?
> 

There are already several places simply dropping the locks and we haven't seen any report about data corruption/inconsistencies/loss caused by it yet.

We have noticed the issue you mentioned and @leonid-s-usov and me agreed we should always drop the locks for the youngest request instead.

> > In my example, the request would have to drop the `xattr` lock and then acquire both the `file` and the `xattr` locks in a single call to maintain the order and prevent potential deadlocks. It's not always possible to do, because dropping xattr lock would break the atomicity of the flow wrt the values protected by the xattr lock, which may or may not be acceptable. If we can't afford it, the solution for that example flow would be to take file lock along with xattr lock from the beginning, for atomicity.
> 
> so lets say a req had a xattr lock and now wants a file locks, so the MDS would drop the xattr lock then acquire the file lock and then xattr lock? and in the meantime what happens if the something else took up xattr?

No, for the next retry of the request the lock order won't change.

Reply:
☝🏻 What Xiubo said!

> Why/how could it be counter-productive? An already known lock set would mean better clarity?

It depends on the particular request's logic. For a complicated process, it may be too expensive to take all locks at once when there are many different paths the logic could take.

> till now (since this patch's need came up recently) I assume things worked fine before?

Violating the lock order doesn't guarantee a deadlock, it just makes it possible. We have very few known deadlock cases (see Xiubo's response for the recent ones), as common ones have already been addressed. Nevertheless, we should know about any requests that break the locking order before we get into more deadlocks by slightly affecting the timings of processes, or adding new requests written correctly but executing at the "wrong" time.

> If we're dropping locks just like this, won't it be disastrous?

This needs to be analyzed and concluded for every given flow, there's no generic response here.

Reply:
@lxbsz 
> We have noticed the issue you mentioned and @leonid-s-usov and me agreed we should always drop the locks for the youngest request instead.
So we're assuming that old requests should've completed their workload/IO? Is this always the case? If only the new requests are taken into consideration, that means we're okay with old requests out-of-order locks?

Reply:
@leonid-s-usov 
> Violating the lock order doesn't guarantee a deadlock, it just makes it possible. We have very few known deadlock cases (see Xiubo's response for the recent ones), as common ones have already been addressed. Nevertheless, we should know about any requests that break the locking order before we get into more deadlocks by slightly affecting the timings of processes, or adding new requests written correctly but executing at the "wrong" time.

So a wrong timing would actually a good timing since it will avoid the deadlocks, is this always guaranteed in distributed systems? Can we always be sure of making the timings as-intended?

Reply:
From the client's perspective, it doesn't make any difference which request got re-scheduled internally by the MDS as long as both requests finish eventually. 

For one, it seems fair to let the older request proceed and the newer request retry.

Secondly, we need to better understand what it means to "retry". It's yet unclear whether all our requests are immune to a sudden replay, but I assume that should be the case given the journaling and request identifiers and other means of recoveries that Ceph is capable of.
If we're unsure, we could additionally argue that there are less chances for the new request to suffer from a retry just because it's probably done less by now, but that's a very shaky argument. However, it supports the approach of picking the newer request for the retry, letting the older proceed as we drop locks.

Finally, we should mention that retry itself is a dangerous endeavor. I don't think we are ready to introduce it, because just dropping the locks and rescheduling may not work for all requests. We need to properly reinitialize internal request state to make sure it genuinely starts anew, and not gets into some unexpected state. I think that the safest way to do this would be to destroy the MDR and re-create it from MClientRequest we are holding on to, as if we just received it from the client. This is only possible for client requests, apparently; I don't think we have a good restart solution for internal requests. With this in mind, it's possible that when we detect a deadlock between a client request and an internal request, we may be forced to retry the client request even if it's older.

Another point to note is that if we go for a retry to avoid a deadlock, we need to be sure it would be a real deadlock, but it requires cross-checking all other MDRs that are waiting for any of the locks we already hold, so a cartesian product of checks.

If we decide to postpone active deadlock evasion, and instead focus on collecting info about **potential** deadlocks, then we can simply warn about flows that take locks out of order, and then decide what to do with those requests.

Reply:
> Finally, we should mention that retry itself is a dangerous endeavor. I don't think we are ready to introduce it, because just dropping the locks and rescheduling may not work for all requests. We need to properly reinitialize internal request state to make sure it genuinely starts anew, avoiding some unexpected states. I think that the safest way to do this would be to destroy the MDR and re-create it from MClientRequest we are holding on to, as if we just received it from the client. This is only possible for client requests, apparently; I don't think we have a good restart solution for internal requests. With this in mind, it's possible that when we detect a deadlock between a client request and an internal request, we may be forced to retry the client request even if it's older.

won't adding more complexity to the flow incur additional turbulence? if a newer req (internally) gets retried internally and if the client retries too, what happens? do we stop client from retrying momentarily?